### PR TITLE
Add Support for Creating Revision Log Messages with Update Tasks

### LINKED ISSUE #95: Provide option to include a revision log message
Some repo admins might want to know that an object was created by Islandora Workbench. This structure works for `POST`:

```json
{
  "type": [
    {
      "target_id": "islandora_object",
      "target_type": "node_type"
    }
  ],
  "title": [
    {
      "value": "Created a collection with POST"
    }
  ],
  "revision_log": [
    {
      "value": "Created by Islandora Workbench."
    }
  ],  
  "field_model": [
    {
      "target_id": 23,
      "target_type": "taxonomy_term"
    }
  ]
}
```

Workbench could provide an option to include "revision_log" in creating or updating object.

### PR DESCRIPTION
**Pull request by hassanelsheikha**
## Link to Github issue or other discussion

Closes #95

## What does this PR do?

This PR adds the ability to create a new revision when running Workbench with the `update` or `update_media` routines. 

## What changes were made?

The JSON for the PATCH requests in the aforementioned routines were modified to include revision log messages. Additionally, as seen in [this PR](https://github.com/mjordan/islandora_workbench_integration/pull/28), new hooks have been created to create the new revision (and include the timestamp in the case of the `update` process).

## How to test / verify this PR?

Integration tests can be performed while monitoring the "Revisions" tab on the media/node.

## Interested Parties

@mjordan 

---

## Checklist

* [x] Before opening this PR, have you opened an issue explaining what you want to to do?
* [x] Have you run `pycodestyle --show-source --show-pep8 --ignore=E402,W504 --max-line-length=200 yourfile.py`? 
* [ ] Have you included same configuration and/or CSV files useful for testing this PR?
* [ ] Have you written unit or integration tests if applicable?
* [ ] Does the code added in this PR require a version of Python that is higher than the current minimum version?
* [ ] If the changes in this PR require an additional Python library, have you included it in `setup.py`?
* [ ] If the changes in this PR add a new configuration option, have you provided a default for when the option is not present in the .yml file?

### CODE REVIEW

**Comment by hassanelsheikha** [comment]
@mjordan Can you please kindly review this PR (and this [prerequisite PR](https://github.com/mjordan/islandora_workbench_integration/pull/28)) when you get the chance? Thank you so much.

**Comment by mjordan** [comment]
Yes, sorry, will do so today.

File: workbench [hunk 1]
```diff
@@ -963,6 +982,36 @@ def update_media() -> None:
                 {'value': status}
             ]
         }
+    
+    def is_revisions_enabled(config: dict, media_type: str) -> bool:
```
Comment (mjordan) [inline workbench:986]: I don't think the type hint `-> bool` is compatible with Workbench's current minimum required version of Python, 3.7. We need to update to 3.8 soon though. Can you remove this for now please?

**Comment by hassanelsheikha** [comment]
@mjordan Do you recall the names of these lists?

**Comment by mjordan** [comment]
The following places:

- https://github.com/mjordan/islandora_workbench/blob/main/workbench_utils.py#L1422
- https://github.com/mjordan/islandora_workbench/blob/main/workbench_utils.py#L1893-L1900

Also, do you know if the database field that stores the revision log entry has a maximum length? Drupal uses a lot of varchar fields in its schema that have a max length of 255 characters. `--check` should also validate the length of the message and warn the user if it exceeds the maximum length allowed. I can add that later if you want though.

**Comment by hassanelsheikha** [comment]
Ok, I think I resolved the `revision_log` issue. 

> Also, do you know if the database field that stores the revision log entry has a maximum length? Drupal uses a lot of varchar fields in its schema that have a max length of 255 characters. `--check` should also validate the length of the message and warn the user if it exceeds the maximum length allowed. I can add that later if you want though.

I'm not sure about this, sorry! Is this pertinent to this PR specifically or would you accept its resolution in a future PR?

**Comment by mjordan** [comment]
Future PR is fine, I can also do it in the meantime.

**Comment by hassanelsheikha** [comment]
I really appreciate it. If there's anything else I need to fix in this PR, I'll be happy to fix it.

**Comment by mjordan** [comment]
Testing with an `update` task, I'm getting the following exception:

```
│    527 │   │   │   content_type_endpoint = f"{config['host']}/entity/node_type/{config['content  │
│    528 │   │   │   response = issue_request(config, 'GET', content_type_endpoint).text           │
│    529 │   │   │   # See if revisions are enabled for the content type and create a new revisio  │
│ ❱  530 │   │   │   revisions_enabled = json.loads(response)['new_revision']                      │
│    531 │   │   │   if revisions_enabled:                                                         │
│    532 │   │   │   │   # Add revision information to node                                        │
│    533 │   │   │   │   revision_log_message = row['revision_log'] if 'revision_log' in row else 
KeyError: 'new_revision'
```

Config is:

```
task: update
host: https://islandora.traefik.me
username: admin
password: password
input_csv: pr_645_update.csv
```
CSV is:
```
node_id,field_description,revision_log
1,Node 1's description.,node 1 updated by mj.
2,Node 2's description.,node 2 updated by mj.
3,Node 3's description.,node 3 updated by mj.
```

**Comment by hassanelsheikha** [comment]
That's strange... Can you please paste the JSON you have at 
https://islandora.traefik.me/entity/node_type/islandora_object/?_format=json

**Comment by mjordan** [comment]
`'No route found for "GET /entity/node_type/islandora_object/" `

I'm using the stock ISLE starter site.

**Comment by hassanelsheikha** [comment]
It's a REST setting. Please try enabling GET for this one

<html>
<body>
<span><span class="ui-provider en bbb bbc bbd bbe bbf bbg bbh bbi bbj bbk bbl bbm bbn bbo bbp bbq bbr bbs bbt bbu bbv bbw bbx bby bbz bca bcb bcc bcd bce bcf bcg bch bci" dir="ltr">

Content type (read-only) | /entity/node_type/{node_type}: GET | methods: GET         formats: jsonld, json, xml authentication: jwt_auth, basic_auth, cookie
-- | -- | --


</span></span>
</body>
</html>

**Comment by mjordan** [comment]
OK, enabling that REST resource worked. Can you add to https://github.com/mjordan/islandora_workbench_integration/pull/28 the following in a file located at `config/optional/rest.resource.entity.node_type.yml` to automatically enable that REST resource?

```yaml
uuid: 25aef6ba-752b-4b62-9140-69043c383363
langcode: en
status: true
dependencies:
  module:
    - basic_auth
    - node
    - serialization
id: entity.node_type
plugin_id: 'entity:node_type'
granularity: resource
configuration:
  methods:
    - GET
  formats:
    - json
  authentication:
    - basic_auth
 ```

**Comment by hassanelsheikha** [comment]
Done!

**Comment by mjordan** [comment]
Just noticed something else. I reran the workbench task and the revision message worked, but the earlier revision isn't showing up in the node's list of revisions. When you update a node via the GUI, earlier revision entries show up in the log:

image

But using Workbench, only the latest shows up:

image

Can  you test to see if multiple revisions of the same node using your PR are showing up for you?

**Comment by hassanelsheikha** [comment]
I tested and I think the issue is with the hook not registering when you change the file. You have to uninstall and reinstall the module for it to work. Please kindly let me know if this works.

**Comment by mjordan** [comment]
Reinstalling the module didn't help I'm afraid. As far as I can tell, the revisions recorded by workbench are replacing the most recent revision, they are not being appended to the list of revisions. I've run workbench a couple times with different field data and revision log messages and the behavior is the same.

**Comment by hassanelsheikha** [comment]
That's very strange. @Natkeeran can you please confirm this worked on your machine? I wonder if there is something different between our installations of ISLE impacting this...

**Comment by mjordan** [comment]
Adding a logging statement:

```php
  // Check if the content type exists and if the "Create new revision" option is enabled.
  if ($content_type && $content_type->shouldCreateNewRevision()) {
    // The "Create new revision" option is enabled.
    \Drupal::logger('islandora_workbench_integration')->warning('islandora_workbench_integration_node_presave is firing');
    $node->setNewRevision(TRUE);
    $node->setRevisionCreationTime(REQUEST_TIME);
  }
```
running `drush pm:uninstall islandora_workbench_integration` then `drush en -y islandora_workbench_integration`, then changing my workbench CSV data and rerunning the task does not yield any messages in my Drupal log.

**Comment by hassanelsheikha** [comment]
Then that hook is never hit. I'm confused now, since that would print a log message for me.

**Comment by mjordan** [comment]
Possibly. I have a Playbook running on another laptop. I'll try on that machine.

**Comment by mjordan** [comment]
Huh, this appears to work as intended on the Playbook, which was originally built in May):

image

However, one thing that I don't see working on either the Playbook or in the ISLE starter site is when the `revision_log` field in the CSV is blank, the default message "" is not used:

image

**Comment by hassanelsheikha** [comment]
Oops, I only handled the case where that column didn't exist at all. Please confirm if the last commit resolved it?

**Comment by hassanelsheikha** [comment]
Also, when testing media please ensure that the column is named "revision_log" and not "revision_log_message". As this PR was a while ago I'm not actually sure why there is a name difference. We can check what happens if I make it "revision_log_message" instead if this difference is not favorable for you.

**Comment by mjordan** [comment]
OK, the last commit provides the default message if the cell is empty.

But, back on my ISLE instance, the revision log entries are still working the way I describe above, not the (intended) way on the older Playbook. On the ISLE instance I reinstalled the integration module and then did a `drush cr` but that didn't have any effect. The only major difference I can think of is that the ISLE version of Drupal is 9.5.9, but the Playbook version is 9.4.3.

**Comment by hassanelsheikha** [comment]
I use an ISLE instance too, but it works fine. Do you have another ISLE instance you can test with?

**Comment by mjordan** [comment]
No, but I could rebuild this one.

**Comment by hassanelsheikha** [comment]
I think that could be our next best choice then. Because I don't think there's any setting other than REST and reinstalling the module that could have caused this.

**Comment by mjordan** [comment]
I rebuilt a fresh ISLE starter site, replaced the Islandora Workbench Integration module in it with yours, same result as yesterday. I think we should be getting someone else to test it. If they are using the same starter site I am, the problem is on my end. If they see the same behavior as me, it must be something else. I can ask on the slack channel for volunteers if you want.

**Comment by hassanelsheikha** [comment]
OK. Please allow me some time; I'll perform some additional tests with other machines and report back to you. I'm really sorry about this, but I'll try my best to get to the bottom of it.

**Comment by mjordan** [comment]
No problem, thanks for looking into it.

**Comment by Natkeeran** [comment]
@mjordan 

I tested it with latest isle-dc and starter.

```
nat@LIB-6HSWM53:~/tools/isle-dc$ git log -1
commit 8b7e11cbc485eaf998d665d1e2ed68c0d4f41be0 (HEAD -> development, origin/development, origin/HEAD)
Merge: 37cb8e7 bbdd83f
Author: aOelschlager <41480572+aOelschlager@users.noreply.github.com>
Date:   Wed Aug 2 12:43:29 2023 -0500

    Merge pull request #338 from joshdentremont/patch-4
    
    fix for public file export path

```

Here are the steps I took.

I uninstalled the islandora_workbench_integration.  Wget https://raw.githubusercontent.com/hassanelsheikha/islandora_workbench_integration/main/islandora_workbench_integration.module.  Changed the ownership to the same owner as other files in that folder.

I enabled content type REST permissions
```
Content type (read-only) | /entity/node_type/{node_type}: GET | methods: GETformats: jsonld, json, xmlauthentication: jwt_auth, basic_auth, cookie
```

Then, I ran the update.  It worked for me as expected (i.e created revisions with appropriate log messages).

Yesterday, I did not uninstall and reinstall the islandora_workbench_integration, and it did not work.  You may want to add some log messages in the hooks to see if it gets there.  If the hooks are not firing, then it won't work as expected.

**Comment by mjordan** [comment]
Thanks @Natkeeran I'll try to replicate your steps. Can you try unchecking the jwt_auth and cookie authentication to see if it still works? None of the other REST endpoints use those, only basic.

**Comment by Natkeeran** [comment]
Unchecked them. Works.

**Comment by mjordan** [comment]
I've followed @Natkeeran's steps (with the addition of re-enbling the module after wgetting Hassan's). I hadn't changed the permissions to nginx previously but that didn't work for me either. I added some logging and it's not showing up in the Drupal reports, so there's something else I'm missing.

That all said, I'm convinced that the problem lies with me. I'm OK merging this, and also merging https://github.com/mjordan/islandora_workbench_integration/pull/28 and cutting a new tag for that module. If there are outstanding issues we can fix them separately. Let me smoketest the media part of this on my playbook and if that goes as planned (it should), I'll merge.

**Comment by hassanelsheikha** [comment]
Sounds good. Thanks a lot!

**Comment by mjordan** [comment]
Checking `update_media` tasks this now in my Playbook instance, where creating node revisions with log entries works. I'm not familiar with media revisions - where should the list of revisions be showing up? I manually updated a media entity and left a log message in the form, but I still don't see the list of revisions.

**Comment by hassanelsheikha** [comment]
Done. Can you please confirm your media update tests still work?

**Pull request merged**